visualizer.py

```diff
@@ -0,0 +1,323 @@
+        while True:
+            await self._update()
+
+
+    def _append_traces(self, frame):
+        # Next, put the trace data into our trace buffers.
+        for trace in frame.to_traces():
+            if trace.label not in self.traces:
+                self.traces[trace.label] = Trace(scale=trace.scale,
+                                                 label=trace.label,
+                                                 values=np.empty(self.max_size))
+
+            num_elements = len(trace.values)
+
+            if self.trace_index + num_elements > len(self.traces[trace.label].values):
+                num_elements = len(self.traces[trace.label].values) - self.trace_index
+
+            self.traces[trace.label].values[self.trace_index:self.trace_index+num_elements] = trace.values[:num_elements]
+
+        # Extend the timebase
+        sample_index = frame.header.sequence * frame.header.batch_size
+        self.times[self.trace_index:self.trace_index+num_elements] = \
+            [np.uint32(sample_index + offset & 0xFFFFFFFF) for offset in range(num_elements)]
+
+        self.trace_index += num_elements
+
+
+    async def _update(self):
+        """ Ingest a single livestream frame. """
+        # Ingest stream frames.
+        frame = await self._stream.queue.get()
+
+        if self.trigger is not TriggerState.TRIGGERED:
+            return
+
+        if frame.format_id != self._current_format:
+            self.traces = dict()
+            self.times = np.empty(self.max_size, dtype=np.uint32)
+            self._current_format = frame.format_id
+
+        # Append trace data
+        self._append_traces(frame)
+
+        # Check for the capture buffer being filled.
+        if self.trace_index >= self.max_size:
+            logging.info('Got %d samples - stopping trigger', self.max_size)
+            self.trigger = TriggerState.STOPPED
+
+
+    def get_traces(self):
+        """ Get the traces and times to display in the graph. """
+        return (self.times - self.times[0]), self.traces
+
+
+class StreamVisualizer:
+    """ A visualizer for stabilizer's livestream data. """
+
+    def __init__(self, receiver: StreamReceiver, document: bokeh.document.Document):
+        """ Initialize the visualizer.
+
+        Args:
+            receiver: The StreamReceiver object for managing stream reception.
+            document: The Bokeh document to draw into.
+        """
+
+        self.receiver = receiver
+
+        figure = bokeh.plotting.figure(output_backend="webgl", sizing_mode='stretch_both')
+
+        # Add a trigger button
+        trigger_button = bokeh.models.Button(label='Single', button_type='primary')
+        trigger_button.on_click(self.capture)
+
+        force_button = bokeh.models.Button(label='Force', button_type='primary')
+        force_button.on_click(self.receiver.force_trigger)
+
+        self._capture_duration_input = bokeh.models.TextInput(title='Capture Duration',
```

Comment:
Can this be tied to the x axis range of the viewport?

Reply:
Hm.... So instead of a textinput, we just request the current zoom of the graph front-end? That's not a bad idea. I'll look into how feasible that is. It would certainly be nicer.